lib/ansible/modules/monitoring/datadog_downtime.py

```diff
@@ -0,0 +1,205 @@
+            api_key=dict(required=True, no_log=True),
+            app_key=dict(required=True, no_log=True),
+            state=dict(required=True, choises=['scheduled', 'canceled']),
+            message=dict(required=False, default=None),
+            scope=dict(required=True),
+            start=dict(required=False, default=None, type='int'),
+            end=dict(required=False, default=None, type='int'),
+            id=dict(required=False)
+        )
+    )
+
+    # Prepare Datadog
+    if not HAS_DATADOG:
+        module.fail_json(msg='datadogpy required for this module')
+
+    options = {
+        'api_key': module.params['api_key'],
+        'app_key': module.params['app_key']
+    }
+
+    initialize(**options)
+
+    if module.params['state'] == 'scheduled':
+        schedule_downtime(module)
+    elif module.params['state'] == 'canceled':
+        cancel_downtime(module)
+
+def _fix_template_vars(message):
+    if message:
+        return message.replace('[[', '{{').replace(']]', '}}')
+    return message
+
+def _get_downtime(module):
+    if module.params['id'] is not None:
+        downtime = api.Downtime.get(module.params['id'])
+        if 'errors' in downtime:
+            module.fail_json(msg="Failed to retrieve downtime with id %s, errors are %s" % (module.params['id'], str(downtime['errors'])))
+        return downtime
+    return {}
+
+def _post_downtime(module):
+    try:
+        kwargs = dict(message=_fix_template_vars(module.params['message']), scope=module.params['scope'],
+            start=module.params['start'], end=module.params['end'])
+        msg=api.Downtime.create(**kwargs)
+        if 'errors' in msg:
+            module.fail_json(msg=str(msg['errors']))
+        else:
+            module.exit_json(changed=True, msg=msg)
+    except Exception:
+        e = get_exception()
+        module.fail_json(msg=str(e))
+
+def _equal_dicts(a, b, ignore_keys):
+    ka = set(a).difference(ignore_keys)
+    kb = set(b).difference(ignore_keys)
+    return ka == kb and all(a[k] == b[k] for k in ka)
+
+def _update_downtime(module, downtime):
+    try:
+        kwargs = dict(id=downtime['id'], message=_fix_template_vars(module.params['message']),
+                      scope=module.params['scope'], start=module.params['start'], end=module.params['end'])
+        msg=api.Downtime.update(**kwargs)
+        if 'errors' in msg:
+            module.fail_json(msg=str(msg['errors']))
+        elif _equal_dicts(msg, downtime, ['creator_id', 'updater_id']):
+            module.exit_json(changed=False, msg=msg)
+        else:
+            module.exit_json(changed=True, msg=msg)
+    except Exception:
+        e = get_exception()
+        module.fail_json(msg=str(e))
+
+def schedule_downtime(module):
+    downtime = _get_downtime(module)
+    if not downtime:
+        _post_downtime(module)
+    else:
+        _update_downtime(module, downtime)
+
+def cancel_downtime(module):
+    downtime = _get_downtime(module)
+    if not downtime:
```

Comment:
This doesn't allow the caller to cancel downtime by scope. _get_downtime returns an empty dictionary when id is not specified. It'd be great if this was changed to require id or scope instead of just id.